app/src/main/java/com/evanfuhr/pokemondatabase/utils/EvolutionDialog.java

```diff
@@ -0,0 +1,168 @@
+        }
+
+        for (Map.Entry<Evolution.Detail, Integer> entry : mEvolution.getTriggerDetails().entrySet()) {
+            switch (entry.getKey()) {
+                case MINIMUM_LEVEL:
+                    prose += ", starting at Lv. " + entry.getValue();
+                    break;
+                case TIME_OF_DAY:
+                    prose += ", only during the " + Evolution.Detail.getTimeOfDay(entry.getValue());
+                    break;
+                case GENDER_ID:
+                    prose += ", if a " + Evolution.Detail.getGender(entry.getValue());
+                case MINIMUM_HAPPINESS:
+                    // If this field is present, it's high happiness, e.g. 220
+                    prose += ", with high friendship";
+                    break;
+                case MINIMUM_AFFECTION:
+                    prose += ", with at least " + entry.getValue() + " affection hearts";
+                    break;
+                case LOCATION_ID:
+                    prose += ", only at " + locationDAO.getLocation(new Location(entry.getValue())).getName();
+                    break;
+                case KNOWN_MOVE_ID:
+                    prose += ", if it knows " + moveDAO.getMove(new Move(entry.getValue())).getName();
+                    break;
+                case KNOWN_MOVE_TYPE:
+                    prose += ", if it knows a " + typeDAO.getType(new Type(entry.getValue())).getName() + " type move";
+                    break;
+                case HELD_ITEM_ID:
+                    prose += ", while holding the " + itemDAO.getItem(new Item(entry.getValue())).getName();
+                    break;
+                case TRIGGER_ITEM_ID:
+                    // This is handled above. It makes the most sense to preserve readability for the end user
+                    break;
+                case PARTY_TYPE_ID:
+                    prose += ", if there is a " + typeDAO.getType(new Type(entry.getValue())).getName() +
+                            " type Pokémon also in the party";
+                    break;
+                case PARTY_SPECIES_ID:
+                    prose += ", if there is a " + pokemonDAO.getPokemon(new Pokemon(entry.getValue())).getName() +
+                            " also in the party";
+                    break;
+                case TRADE_SPECIES_ID:
+                    prose += ", if traded for a " + pokemonDAO.getPokemon(new Pokemon(entry.getValue())).getName();
+                    break;
+                case NEEDS_OVERWORLD_RAIN:
+                    prose += ", if it's raining outside of battle";
+                    break;
+                case TURN_UPSIDE_DOWN:
+                    prose += ", while holding the device upside down";
+                    break;
+                case MINIMUM_BEAUTY:
+                    prose += ", with a maxed out beauty stat";
+                    break;
+                case RELATIVE_PHYSICAL_STATS:
+                    switch (entry.getValue()) {
```

Comment:
Wow double switch... all the way! i've never seen such structure before but looks clean.

Consider breaking up this method somehow. It's a bit large, but if it makes sense, leave it.

Reply:
I could turn it into an enum, but that seems unnecessary given the content that needs to be generated. How about instead of breaking up the method, I add better (read: actual) documentation?

Reply:
I agree, documentation is fine then.